src/tracing.rs

```diff
@@ -67,6 +46,466 @@ pub fn setup(user_filters: &str) -> anyhow::Result<()> {
     ));
 
     tracing::subscriber::set_global_default(reg)?;
+    tracing_log::LogTracer::init()?;
+
+    Ok(logger_handle)
+}
+
+#[derive(Clone)]
+pub struct LoggerHandle {
+    config: Arc<RwLock<config::LoggerConfig>>,
+    default: DefaultLoggerReloadHandle,
+}
+
+#[rustfmt::skip] // `rustfmt` formats this into unreadable single line
+type DefaultLoggerReloadHandle<S = Registry> = reload::Handle<
+    default::Logger<S>,
+    S,
+>;
+
+impl LoggerHandle {
+    pub fn new(config: config::LoggerConfig, default: DefaultLoggerReloadHandle) -> Self {
+        Self {
+            config: Arc::new(RwLock::new(config)),
+            default,
+        }
+    }
+
+    pub async fn get_config(&self) -> config::LoggerConfig {
+        self.config.read().await.clone()
+    }
+
+    pub async fn update_config(&self, diff: config::LoggerConfigDiff) -> anyhow::Result<()> {
+        let mut config = self.config.write().await;
+
+        config.default.update(diff.default);
+        let default = default::new(&config.default);
+        self.default.reload(default)?;
+
+        Ok(())
+    }
+}
+
+pub mod config {
+    use super::*;
+
+    #[derive(Clone, Debug, Default, Eq, PartialEq, Deserialize, Serialize)]
+    #[serde(default)]
+    pub struct LoggerConfig {
+        #[serde(flatten)]
+        pub default: default::Config,
+    }
+
+    impl LoggerConfig {
+        pub fn with_top_level_directive(&mut self, log_level: Option<String>) -> &mut Self {
+            if self.default.log_level.is_some() && log_level.is_some() {
+                eprintln!(
+                    "Both top-level `log_level` and `logger.log_level` config directives are used. \
+                     `logger.log_level` takes priority, so top-level `log_level` will be ignored."
+                );
```

Comment:
I assume we're using `eprintln!` here because our logging infrastructure isn't ready here yet?

Reply:
Yep.